paddle/math/float16.h

```diff
@@ -0,0 +1,714 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <cstdint>
+
+#include <cuda.h>
+#include "unsupported/Eigen/CXX11/Tensor"
+
+#ifdef __GNUC__
+#define PADDLE_GNUC_VER (__GNUC__ * 10 + __GNUC_MINOR__)
+#else
+#define PADDLE_GNUC_VER 0
+#endif  // __GNUC__
+
+#ifdef __clang__
+#define PADDLE_CLANG_VER (__clang_major__ * 10 + __clang_minor__)
+#else
+#define PADDLE_CLANG_VER 0
+#endif  // __clang__
+
+#ifdef __CUDACC__
+#define PADDLE_HOSTDEVICE __host__ __device__
+#if CUDA_VERSION >= 7050
+#define PADDLE_CUDA_FP16
+#include <cuda_fp16.h>
+#endif  // CUDA_VERSION >= 7050
+#else
+#define PADDLE_HOSTDEVICE
+#endif  // __CUDACC__
+
+#ifdef __arm__
+#define PADDLE_ARM_32
+#endif
+
+#ifdef __aarch64__
+#define PADDLE_ARM_64
+#endif
+
+#if defined(PADDLE_ARM_32) || defined(PADDLE_ARM_64)
+#define PADDLE_ARM
+#endif
+
+#if defined(__ARM_NEON) || defined(__ARM_NEON__)
+#define PADDLE_NEON
+#include <arm_neon.h>
+#endif
+
+#if defined(PADDLE_NEON) && defined(PADDLE_ARM_32)
+#define PADDLE_NEON_32
+#endif
+
+#if defined(PADDLE_NEON) && defined(PADDLE_ARM_64)
+#define PADDLE_NEON_64
+#endif
+
+#ifdef PADDLE_ARM
+#ifdef __F16C__
+#undef __F16C__
+#endif  // __F16C__
+#else
+#include <immintrin.h>
+#endif  // PADDLE_ARM
+
+#define PADDLE_ALIGN(x) __attribute__((aligned(x)))
+
+namespace paddle {
+
+struct float16;
+
+namespace fp16_impl {
+// Convert from float to half precision in round-to-nearest-even mode
+PADDLE_HOSTDEVICE inline float16 float_to_half_rn(float f);
+PADDLE_HOSTDEVICE inline float half_to_float(float16 h);
+}  // namespace fp16_impl
+
+// Use PADDLE_ALIGNED(2) to ensure that each float16 will be allocated
+// and aligned at least on a 2-byte boundary, which leads to efficient
+// memory access of float16 struct and also makes float16 compatible
+// with CUDA half, ARM float16_t, and Eigen::half data types.
+struct PADDLE_ALIGN(2) float16 {
+  uint16_t x;
+
+  PADDLE_HOSTDEVICE inline float16() : x(0) {}
+
+  PADDLE_HOSTDEVICE inline float16(const float16& h) : x(h.x) {}
+
+#ifdef PADDLE_CUDA_FP16
+  PADDLE_HOSTDEVICE inline float16(const half& h) {
+#if CUDA_VERSION >= 9000
+    x = reinterpret_cast<__half_raw*>(&h)->x;
+#else
+    x = h.x;
+#endif  // CUDA_VERSION >= 9000
+  }
+#endif  // PADDLE_CUDA_FP16
+
+  PADDLE_HOSTDEVICE inline float16(const Eigen::half& h) : x(h.x) {}
+
+#if defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
```

Comment:
Where is `PADDLE_ARM_FP16` define?

Reply:
We can define `-DPADDLE_ARM_FP16` in cmake when the architecture is specified as ARMv8.2